metrics-core/src/main/java/com/codahale/metrics/SlidingTimeWindowArrayReservoir.java

```diff
@@ -0,0 +1,98 @@
+package com.codahale.metrics;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A {@link Reservoir} implementation backed by a sliding window that stores only the measurements made
+ * in the last {@code N} seconds (or other time unit).
+ */
+public class SlidingTimeWindowArrayReservoir implements Reservoir {
+    // allow for this many duplicate ticks before overwriting measurements
+    private static final long COLLISION_BUFFER = 256L;
+    // only trim on updating once every N
+    private static final long TRIM_THRESHOLD = 256L;
+    private static final long CLEAR_BUFFER = TimeUnit.HOURS.toNanos(1) * COLLISION_BUFFER;
+
+    private final Clock clock;
+    private final ChunkedAssociativeLongArray measurements;
+    private final long window;
+    private final AtomicLong lastTick;
+    private final AtomicLong count;
+
+    /**
+     * Creates a new {@link SlidingTimeWindowArrayReservoir} with the given window of time.
+     *
+     * @param window     the window of time
+     * @param windowUnit the unit of {@code window}
+     */
+    public SlidingTimeWindowArrayReservoir(long window, TimeUnit windowUnit) {
+        this(window, windowUnit, Clock.defaultClock());
+    }
+
+    /**
+     * Creates a new {@link SlidingTimeWindowArrayReservoir} with the given clock and window of time.
+     *
+     * @param window     the window of time
+     * @param windowUnit the unit of {@code window}
+     * @param clock      the {@link Clock} to use
+     */
+    public SlidingTimeWindowArrayReservoir(long window, TimeUnit windowUnit, Clock clock) {
+        this.clock = clock;
+        this.measurements = new ChunkedAssociativeLongArray();
+        this.window = windowUnit.toNanos(window) * COLLISION_BUFFER;
+        this.lastTick = new AtomicLong(clock.getTick() * COLLISION_BUFFER);
+        this.count = new AtomicLong();
+    }
+
+    @Override
+    public int size() {
+        trim();
+        return measurements.size();
+    }
+
+    @Override
+    public void update(long value) {
+        long newTick;
+        do {
+            if (count.incrementAndGet() % TRIM_THRESHOLD == 0L) {
+                trim();
+            }
+            long lastTick = this.lastTick.get();
+            newTick = getTick();
+            boolean longOverflow = newTick < lastTick;
+            if (longOverflow) {
+                measurements.clear();
```

Comment:
Previous implementation was able to handle nanotime overflow without loosing anything.